File: rx-codelab/src/test/java/com/jraska/rx/codelab/solution/Solution_Task1_Basics.java
package com.jraska.rx.codelab.solution;

import org.junit.Test;

import io.reactivex.Observable;

public class Solution_Task1_Basics {
  @Test
  public void dummyObservable() {
    Observable<String> stringObservable = Observable.just("Hello Rx");

    stringObservable.subscribe(System.out::println);
  }

  @Test
  public void methodIntoObservable() {
    Observable<Long> currentTimeObservable = Observable.fromCallable(System::currentTimeMillis);

    currentTimeObservable.subscribe(System.out::println);
  }

  @Test
  public void helloOperator() {
    Observable<Integer> rangeObservable = Observable.range(1, 10);

    rangeObservable.filter(Solution_Task1_Basics::isOdd)
      .subscribe(System.out::println);
  }

  @Test
  public void receivingError() {
    Observable<Integer> integerObservable = Observable.error(new RuntimeException("I want to crash you"));

    integerObservable.subscribe(System.out::println, System.err::println);
  }

  static boolean isOdd(int value) {
    return value % 2 == 1;
  }
}

File: rx-codelab/src/test/java/com/jraska/rx/codelab/solution/SolutionTask1Basics.kt
package com.jraska.rx.codelab.solution

import io.reactivex.Observable
import org.junit.Test

class SolutionTask1Basics {
  @Test
  fun dummyObservable() {
    val stringObservable = Observable.just("Hello Rx")

    stringObservable.subscribe { println(it) }
  }

  @Test
  fun methodIntoObservable() {
    val currentTimeObservable = Observable.fromCallable<Long> { System.currentTimeMillis() }

    currentTimeObservable.subscribe { println(it) }
  }

  @Test
  fun helloOperator() {
    val rangeObservable = Observable.range(1, 10)

    rangeObservable.filter { isOdd(it) }
      .subscribe(System.out::println)
  }

  @Test
  fun receivingError() {
    val integerObservable = Observable.error<Int>(RuntimeException("I want to crash you"))

    integerObservable.subscribe(::println, System.err::println)
  }

  companion object {
    fun isOdd(value: Int): Boolean {
      return value % 2 == 1
    }
  }
}
